miasm2/ir/ssa.py

```diff
@@ -339,3 +339,267 @@ def transform(self, path):
         """
         for block in path:
             self._rename_expressions(block)
+
+
+class SSADiGraph(SSA):
+    """
+    SSA transformation on DiGraph level
+
+    It handles
+    - transformation of a DiGraph into SSA
+    - generation, insertion and filling of phi nodes
+
+    The implemented SSA form is known as minimal SSA.
+    """
+
+    def __init__(self, ira):
+        """
+        Initialises SSA class for directed acyclic graphs
+        :param ira: instance of IRA
+        """
+        super(SSADiGraph, self).__init__(ira)
+
+        # variable definitions
+        self.defs = {}
+
+        # dict of blocks' phi nodes
+        self._phinodes = dict()
+
+        # IRA control flow graph
+        self.graph = ira.graph
+
+    def transform(self, head):
+        """Transforms into SSA"""
+        self._init_variable_defs(head)
+        self._place_phi(head)
+        self._rename(head)
+        self._insert_phi()
+        self._convert_phi()
+
+    def reset(self):
+        """Resets SSA transformation"""
+        super(SSADiGraph, self).reset()
+        self.defs = {}
+        self._phinodes = dict()
+
+    def _init_variable_defs(self, head):
+        """
+        Initialises all variable definitions and
+        assigns the corresponding IRA blocks.
+
+        All variable definitions in self.defs contain
+        a set of IRA blocks in which the variable gets assigned
+        """
+        # architecture's instruction pointer
+        instruction_pointer = set(self.ira.arch.pc.values() + [self.ira.IRDst])
+
+        for block in self.graph.walk_depth_first_forward(head):
+            ib = self.get_block(block)
+            # blocks IR expressions
+            ir_expressions = (m2_expr.ExprAff(dst, assignblk[dst])
+                              for assignblk in ib.irs for dst in assignblk)
+            for e in ir_expressions:
```

Comment:
You only use `e.dst` here, that is to say `assignblk` destinations, so you can directly iterate on `assignblk` (which iterate over destinations)